fendermint/vm/interpreter/src/fvm/topdown.rs

```diff
@@ -45,7 +48,14 @@ where
     DB: Blockstore + Sync + Send + 'static,
 {
     let total_value: TokenAmount = messages.iter().map(|a| a.msg.value.clone()).sum();
-    gateway_caller.mint_to_gateway(state, total_value)?;
+
+    gateway_caller
+        .mint_to_gateway(state, total_value.clone())
+        .context("failed to mint to gateway")?;
+
+    state.update_circ_supply(|circ_supply| {
+        *circ_supply += total_value;
+    });
 
     gateway_caller.apply_cross_messages(state, messages)
```

Comment:
I think we already discussed this when we were reviewing top-down, and we said we would defer it to the future, but what is your take for when executing the message fails?

Reply:
I had to re-read my comments:
1. [For crashing](https://github.com/consensus-shipyard/fendermint/pull/310#discussion_r1352791877)
2. [Against crashing](https://github.com/consensus-shipyard/fendermint/pull/310#discussion_r1352803799)

To reiterate the main points:
* Individual messages in the batch should be allowed to fail without failing the whole batch
* Failing the whole batch and moving on is unacceptable, especially if we can't roll back the validator changes and the parent height finality
* But, the whole match might fail if it goes over the block gas limit
* Therefore we should:
  * do all 3 parts as an atomic call to Solidity
  * let the whole thing fail as a transaction and be retried by another relayer
  * not execute immediately (and hit the block gas limit) but enqueue for execution and do it when we have room
  * then we can treat individual messages with their own gas limit